subprojects/core/src/main/java/org/gradle/process/internal/worker/request/Receiver.java

```diff
@@ -16,25 +16,41 @@
 
 package org.gradle.process.internal.worker.request;
 
+import org.gradle.api.problems.Problem;
 import org.gradle.internal.UncheckedException;
 import org.gradle.internal.dispatch.StreamCompletion;
+import org.gradle.internal.logging.events.LogEvent;
 import org.gradle.internal.logging.events.OutputEventListener;
+import org.gradle.internal.logging.events.StyledTextOutputEvent;
 import org.gradle.internal.remote.internal.hub.StreamFailureHandler;
 import org.gradle.process.internal.worker.DefaultWorkerLoggingProtocol;
+import org.gradle.process.internal.worker.DefaultWorkerProblemProtocol;
 import org.gradle.process.internal.worker.WorkerProcessException;
+import org.gradle.process.internal.worker.child.WorkerLoggingProtocol;
+import org.gradle.process.internal.worker.problem.WorkerProblemProtocol;
 
 import java.util.concurrent.ArrayBlockingQueue;
 import java.util.concurrent.BlockingQueue;
 
-public class Receiver extends DefaultWorkerLoggingProtocol implements ResponseProtocol, StreamCompletion, StreamFailureHandler {
+/**
+ * Receives and handles messages about a given worker action executed by a worker process.
+ * <p>
+ * This receiver is used per worker action.
+ */
+public class Receiver implements ResponseProtocol, StreamCompletion, StreamFailureHandler {
     private static final Object NULL = new Object();
     private static final Object END = new Object();
     private final BlockingQueue<Object> received = new ArrayBlockingQueue<Object>(10);
     private final String baseName;
     private Object next;
 
+    // Sub-handlers for the different protocols implemented by ResponseProtocol
+    private final WorkerLoggingProtocol loggingProtocol;
+    private final WorkerProblemProtocol problemProtocol;
+
     public Receiver(String baseName, OutputEventListener outputEventListener) {
-        super(outputEventListener);
+        this.loggingProtocol = new DefaultWorkerLoggingProtocol(outputEventListener);
+        this.problemProtocol = new DefaultWorkerProblemProtocol();
```

Comment:
Originally the logging was implemented by inheritance (this class extended the default logging implementation)
This didn't feel right after adding the problem reporting and turning this into a composition-based solution instead of inheritance-based one.

Reply:
💭 I would even argue that this could be _all_ composition.  i.e. a separate delegate for all of the functions of `ResponseProtocol` so that this class is really just a fanout for those functions.  I don't think you need to change this (unless you want to) - I'm just sharing the thought that it might make it more clear that the aggregation of these things in `ResponseProtocol` is an implementation detail rather than any true relationship between the functions.

Reply:
I originally wanted to do more but scaled back my efforts; I was afraid a larger change would delay this feature, which we really want to be available in 8.6.